lib/TPP/TransformUtils.cpp

```diff
@@ -208,13 +208,83 @@ bool isBlockedConvolution(Operation *op) {
   return tpp::utils::hasMatmulBody(linalgOp);
 }
 
-// TODO: Check indexing maps and iterator types. They should
-// match the one of a packed matmul.
+static llvm::SmallDenseSet<unsigned> getPreservedDims(AffineMap map) {
+  assert(map.isProjectedPermutation() &&
+         "expected map to have projected permutations");
+  llvm::SmallDenseSet<unsigned> preservedDims;
+  for (auto expr : map.getResults())
+    preservedDims.insert(expr.cast<AffineDimExpr>().getPosition());
+  return preservedDims;
+}
+
 bool isBlockedMatmul(Operation *op) {
-  if (!isa<linalg::LinalgOp>(op))
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp)
     return false;
-  linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(op);
-  return tpp::utils::hasMatmulBody(linalgOp);
+  if (linalgOp.getNumDpsInputs() != 2 || linalgOp.getNumDpsInits() != 1)
+    return false;
+
+  // Check the input indexing map has the right form.
+  auto indexingMaps = linalgOp.getIndexingMapsArray();
+  if (indexingMaps.size() != 3)
+    return false;
+  if (llvm::any_of(indexingMaps, [](AffineMap map) {
+        return !map.isProjectedPermutation();
+      })) {
+    return false;
+  }
+  for (auto map : indexingMaps)
+    for (auto expr : map.getResults())
+      if (!expr.isa<AffineDimExpr>())
+        return false;
+
+  auto iteratorTypes = linalgOp.getIteratorTypesArray();
+
+  // Make sure all loops are characterized as one of:
+  // - I loop: present in C and A but not in B. I must be parallel.
+  // - J loop: present in C and B but not in A. J must be paralell.
+  // - K loop: present in A and B but not in C. K must be reduction.
+
+  llvm::SmallDenseSet<unsigned> cDims = getPreservedDims(indexingMaps.back());
+  llvm::SmallDenseSet<unsigned> bDims = getPreservedDims(indexingMaps[1]);
+  llvm::SmallDenseSet<unsigned> aDims = getPreservedDims(indexingMaps.front());
+  llvm::SmallDenseSet<unsigned> allLoopDims;
+  for (auto cExpr : indexingMaps.back().getResults()) {
+    unsigned cDim = cExpr.cast<AffineDimExpr>().getPosition();
```

Comment:
Too much repetition here, we need to common these up.

We'll soon start looking more and more into maps to match TPP ops, so we need a stable API to get things like number and order of parallel/reduction loops, correct usage of affine map indices, etc.

For now, just commoning these up in a separate helper should do the trick.

Reply:
where is the repetition here? The for loops (i.e., line 252-286)?

Reply:
Yes. There are three identical and one very similar loops here. This looks like an easy refactory.

Reply:
sorry I don't get it. Can you please clarify what you have in mind? Each if uses different sets and check for different conditions.

Reply:
The logic for I,J,K loops are the same, you just have to pass the containers by reference, the dim and iterator type by value, and return boolean. At least these could be a local lambda.

Reply:
Discussed offline, it's fine.